lib/archethic/mining/smart_contract_validation.ex

```diff
@@ -0,0 +1,48 @@
+defmodule Archethic.Mining.SmartContractValidation do
+  @moduledoc """
+  This module provides functions for validating smart contracts remotely.
+  """
+
+  alias Archethic.Election
+  alias Archethic.Contracts
+  alias Archethic.Contracts.Contract
+  alias Archethic.P2P
+  alias Archethic.P2P.Message
+  alias Archethic.TransactionChain
+  alias Archethic.TransactionChain.Transaction
+
+  alias Archethic.TaskSupervisor
+
+  @doc """
+  Fetches contracts from a list of recipients and returns a list of valid contracts.
+  """
+  @spec fetch_contracts(list(binary())) :: Enumerable.t() | list(Contract.t())
+  def fetch_contracts(recipients) do
+    TaskSupervisor
+    |> Task.Supervisor.async_stream_nolink(recipients, &do_fetch_contract/1, timeout: Message.get_max_timeout())
+    |> Stream.filter(&match?({:ok, {:ok, _}}, &1))
+    |> Stream.map(fn {:ok, {:ok, contract}} -> contract end)
+  end
+
+  defp do_fetch_contract(recipient) do
+    storage_nodes = Election.chain_storage_nodes(recipient, P2P.authorized_and_available_nodes())
+
+    case TransactionChain.fetch_transaction_remotely(recipient, storage_nodes) do
+      {:ok, tx = %Transaction{}} ->
+        Contracts.from_transaction(tx)
+
+      {:error, _} = e ->
+        e
+    end
+  end
+
+  @doc """
+  Determine if the smart contracts conditions are valid according to the given transaction
+  """
+  @spec valid_contract_calls?(Enumerable.t() | list(Contract.t()), Transaction.t()) :: boolean()
+  def valid_contract_calls?(contracts, transaction = %Transaction{}) do
+    TaskSupervisor
+    |> Task.Supervisor.async_stream_nolink(contracts, &Contracts.valid_contract_execution?(&1, transaction))
```

Comment:
As discussed with @neylix, we believe we need to fetch the calls of the contract and pass them as 3rd argument.

Reply:
Yes I was thinking about that, I didn't implement it because it will be an expensive operation. We need to figure out a way to assert a transaction without having to pull all the inputs of each destination chain.